### PR TITLE
WIP: Implement h2 protocol

### LINKED ISSUE #47: HTTP 2.0 Support
Using the h2 package.

### PR DESCRIPTION
**Pull request by Vibhu-Agarwal**
Closes #47

### HTTP Parsing
- Using [h2](https://hyper-h2.readthedocs.io/en/stable/index.html)

### PR #214
It was great! I've picked a lot of stuff from there.
I've mostly worked on extending the PR, removing bugs and modernizing the code.
I've opted to omit a few implementations though, on which I'd like to have some discussion.

### Motivation from Hypercorn's code
I also read the [hypercorn](https://pgjones.gitlab.io/hypercorn/)'s code, which has the implementation for HTTP/2 already. I got a pretty good idea of http2's working from there.

#### priority
It uses [priority](https://github.com/python-hyper/priority) to build a priority tree and manage the streams.
I haven't yet implemented this in this PR (as it wasn't, in #214). **Should we use it?**
If we use it, it will also help in implementing code for handling h2 events like `WindowUpdated`, `PriorityUpdated` and `RemoteSettingsChanged` (which are currently ignored).

#### stream_buffers
Hypercorn also uses StreamBuffers to handle the data of different streams per connection. This _may be_ due to the fact that [h2 doesn't buffer data internally for sending](http://python-hyper.org/projects/hyper-h2/en/stable/api.html#h2.connection.H2Connection.send_data). I'm not sure though, as [hypercorn checks the chunk size (before sending) anyway](https://gitlab.com/pgjones/hypercorn/-/blob/0.11.2/src/hypercorn/protocol/h2.py#L151-156).
I haven't yet implemented this in this PR (as it wasn't, in #214). **Should we use it?**

### Server Push
I wasn't sure if I should go ahead with [HTTP Server-Push](https://tools.ietf.org/html/rfc7540#section-8.2)'s implementation or not, as [its status is not clear](https://groups.google.com/a/chromium.org/g/blink-dev/c/K3rYLvmQUBY/m/vOWBKZGoAQAJ).
**Should we support it?**

### In this PR

- Some h2-specific changes in SSL-Context
  - https://github.com/encode/uvicorn/blob/a8dcf147b1044adf7627f03edc5b59dc02cc0fdf/uvicorn/config.py#L215
  - https://github.com/encode/uvicorn/blob/a8dcf147b1044adf7627f03edc5b59dc02cc0fdf/uvicorn/config.py#L116-L125

- No websocket support, *yet*.
  - [Unlike h11](https://h11.readthedocs.io/en/latest/api.html#switching-protocols), a `Connection: Upgrade` (to `webocket`) request raises `h2.exceptions.ProtocolError` here:
    https://github.com/encode/uvicorn/blob/b683626db3affb51cffd54bc809ba3b662a4119d/uvicorn/protocols/http/h2_impl.py#L219-L221
  - This can be resolved through the usage of [Extended CONNECT Method](https://tools.ietf.org/html/rfc8441#page-4).
    https://github.com/encode/uvicorn/blob/b683626db3affb51cffd54bc809ba3b662a4119d/uvicorn/protocols/http/h2_impl.py#L111
    That will have to be handled here:
    https://github.com/encode/uvicorn/blob/b683626db3affb51cffd54bc809ba3b662a4119d/uvicorn/protocols/http/h2_impl.py#L284-L285

### Tests
- For tests for [config.py](https://github.com/Vibhu-Agarwal/uvicorn/blob/h2_impl/uvicorn/config.py),
  - It didn't have much to cover up the coverage, but SSL-things had to be tested. So, [one](https://github.com/Vibhu-Agarwal/uvicorn/blob/df02d8c2f4ce2c9d19c0cb68b2ca614ccf02811a/tests/test_config.py#L169-L178) for that.
- For tests for [h11_impl.py](https://github.com/Vibhu-Agarwal/uvicorn/blob/h2_impl/uvicorn/protocols/http/h11_impl.py),
  - These two upgrades need tests:
    1. h2c-upgrade
    2. H2-Prior-Knowledge Connection-Preface
  - **Work**: I added one for both of them (got the coverage; but checks in response still remain):
    1. h2c-upgrade: I added [one](https://github.com/Vibhu-Agarwal/uvicorn/blob/df02d8c2f4ce2c9d19c0cb68b2ca614ccf02811a/tests/protocols/test_http.py#L76-L85) for a simple GET request.
        For some reason, `test_h2c_upgrade_request` is causing failure of some un-related tests in [Python3.8](https://github.com/encode/uvicorn/runs/2720497221) and [Python3.9](https://github.com/encode/uvicorn/runs/2720497495) (runs fine in [Python <=3.7](https://github.com/encode/uvicorn/runs/2720497174)). Commit 81f21be
        **Manual-Testing**: I'm not sure how to manually test this one.
    2. H2-Prior-Knowledge Connection-Preface: The [test](https://github.com/Vibhu-Agarwal/uvicorn/blob/df02d8c2f4ce2c9d19c0cb68b2ca614ccf02811a/tests/protocols/test_http.py#L87-L93) is really a minimal one, not even hitting any endpoint. So I'm not even sure if this is sufficient or not.
        **Manual-Testing**: I tried this with a POST request using curl. It works. I'm not sure how to replicate it in Python or even get raw HTTP request out of it.
  - Basically, I'm out of ideas to implement tests for these.
- For tests for [h2_impl.py](https://github.com/Vibhu-Agarwal/uvicorn/blob/h2_impl/uvicorn/protocols/http/h2_impl.py),
  - Just like `HTTPToolsProtocol` and `H11Protocol`, `H2Protocol` needs testing for a lot of stuff.
  - I saw two options of implementation methods:
    1. [test_http.py](https://github.com/encode/uvicorn/blob/master/tests/protocols/test_http.py) way (using mocks). I tried this out (had to modify [`get_connected_protocol()`](https://github.com/encode/uvicorn/blob/87da6cf4082cd28306bdb78d7d42552d131cc179/tests/protocols/test_http.py#L161-L170) a bit) but got messed up with valid raw HTTP/2 requests and SSL-Context passing. Couldn't debug and get it to work.
    2. [test_ssl.py](https://github.com/encode/uvicorn/blob/master/tests/test_ssl.py) way (using httpx). I installed Starlette, added a (very) small app, and tried out a test with a simple `POST` request. It kinda worked out. So I've added this one in the PR.
  - **Manual-Testing**: I tried with my (Firefox) browser using a _sane_ app on simple GET requests. It seems to work fine.
    I haven't tried broken apps and others with exceptions.

### WIP
- I've marked a lot of places in the code with `# TODO: Task` and added some comments for decision-related discussion or missing implementation.
- I've also tried to keep the code of `h2_impl.py` similar to `httptools_impl.py` and `h11_impl.py`.

### Help Needed
I'm just giving it a go. Hope to complete this soon 🤞
Will need reviews, inputs, tests, ideas and a lot of help from everyone 🤗

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ def get_packages(package):
 minimal_requirements = [
     "click==7.*",
     "h11>=0.8",
+    "h2>=4.0.0",
```
Comment (Kludex) [inline setup.py:49]: I think @tomchristie expressed a wish for this be an optional requirement, but I can't find the issue. 

Just pinging to confirm :kissing:

**Comment by Vibhu-Agarwal** [comment]
^ So that some discussion could be started.
The Work is still In-Progress though.

File: requirements.txt [hunk 2]
```diff
@@ -20,6 +20,7 @@ types-pyyaml
 trustme
 cryptography
 coverage
+starlette
```
Comment (Kludex) [inline requirements.txt:23]: Do we need `starlette`? What's the idea here?

File: setup.py [hunk 3]
```diff
@@ -47,6 +47,7 @@ def get_packages(package):
     "asgiref>=3.3.4",
     "click>=7.*",
     "h11>=0.8",
+    "h2>=4.0.0",
```
Comment (Kludex) [inline setup.py:50]: We want to do the same as httpx: https://github.com/encode/httpx/blob/ab64f7c41fc0fbe638dd586fecf0689c847109bb/setup.py#L66

File: uvicorn/protocols/http/h2_impl.py [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+                    body = message.get("body", b"")
+                self.conn.send_data(self.stream_id, body, end_stream=(not more_body))
+                self.transport.write(self.conn.data_to_send())
+
+                # Handle response completion
+                if not more_body:
+                    self.response_complete = True
+            elif message_type == "http.response.push":
+                # TODO: Implement or Not?
+                # https://groups.google.com/a/chromium.org/g/blink-dev/c/K3rYLvmQUBY/m/vOWBKZGoAQAJ 😕  # noqa: E501
```
Comment (Kludex) [inline uvicorn/protocols/http/h2_impl.py:556]: We should, yep.

**Comment by lovelydinosaur** [comment]
So, I'm neither necessarily *for* or *against* this, but I do have an alternate perspective that I think is worth sharing/considering.

See the discussion in... https://github.com/encode/uvicorn/discussions/1105

**Comment by lovelydinosaur** [comment]
It may well make sense for us to *deliberately* limit the scope of Uvicorn to being an HTTP/1.1 server.
The tighter the scope, the more focused our resources can be on making sure we do a fantastic job in a particular space.

**Comment by Kludex** [comment]
@Vibhu-Agarwal Thanks for the effort here, but I think by #1105 we'll highlight hypercorn as the recommendation for HTTP/2.

Sorry for taking so long to act here.

**Pull request closed without merging**